sys: clock: add sys_clock api and remove posix from iso c time

cfriedt opened the pull request:
Previously, both the common libc `time()` as well as picolibc `time()` (an ISO C function) relied on either `clock_gettime()` or `gettimeofday()`. That had the unfortunate consequence that many areas had an indirect dependency on POSIX. 

This change removes the dependency cycle that existed between POSIX and some libc's by creating a mutual dependency (i.e. new API) for both POSIX and the common C library under `lib/os`. With this change, the ISO C function `time()` can also be used without relying on POSIX support.

* `sys_clock_gettime()`
* `sys_clock_settime()`
* `sys_clock_nanosleep()`
* `sys_clock_getrtoffset()`

Areas that should no longer have POSIX dependencies for getting the "wall clock" time include, but are not limited to
* libc
  * `lib/libc/common/source/thrd`
  * `lib/libc/common/source/time`
  * picolibc (specifically with `time()`)
* `samples/net/cloud/aws_iot_mqtt`
* `samples/net/lwm2m_client`
* `subsys/logging/log_core.c`
* `subsys/net/lib/config` (sntp)
* `subsys/net/lib/lwm2m`

Please see individual commits for details

* scripts: ci: check_compliance: allow CONFIG_BOARD_UNIT_TESTING
* posix: cond + mutex: log messages generate warnings for 32-bit timeouts
* sys: clock: additional sys_clock api calls
* doc: release: 4.2.0: add sys_clock gettime settime nanosleep notes
* logging: use sys_clock_gettime()
* libc: common: time: use sys_clock api rather than posix
* libc: use the common libc time() implementation for most libcs
* libc: common: thrd: use sys_clock_nanosleep() instead of nanosleep()
* tests: libc: thrd: compare with thrd_success rather than ok or zero
* tests: libc: thrd: do not pass NULL for thrd_sleep() duration
* tests: libc: thrd: use timespec_from_timeout()
* tests: libc: thrd: use sys_clock_gettime() and timespec_add()
* tests: lib: move time testsuite to c_lib
* posix: use sys_clock implementation
* net: remove dependency on posix for iso c time() function
* samples: net: remove POSIX_TIMERS and XSI_SINGLE_PROCESS
* sys: clock: merge sys_clock.h into sys/clock.h
 
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/88555
Fixes #88882

> [!NOTE]  
> There have been many RFCs on this topic already #19030, #40099, #76335 and most of them agree on the need for an abstract clock source at the OS level. There isn't a huge need for another RFC here.

* this PR is fixing a bug at this point, but otherwise should facilitate the improvements mentioned in previous RFCs
* rather than the proposed `K_` prefix, the `SYS_` prefix was suggested by @peter-mitsis 
* there is still a distinct need to communicate clock resolution in a better way (does not need to happen in this PR)
* suggestion for above is to use a dt-chosen node that simply provides metadata for the system uptime counter (resolution, accuracy, constness) and then wrap it in `sys_clock_getres()`.

Commit:
sys: clock: additional sys_clock api calls

Additional entries for the sys_clock API, comprised of:

* sys_clock_gettime()
* sys_clock_settime()
* sys_clock_nanosleep()

along with the constants

* SYS_CLOCK_REALTIME
* SYS_CLOCK_MONOTONIC
* SYS_TIMER_ABSTIME

The primary motivation for this API is so that libc and other
libraries have a familiar-enough API to reach to when POSIX
is not available, since POSIX is optional in Zephyr.

By adding this API to lib/os, we also eliminate dependency cycles
between libc and posix, as lib/os is a mutual dependency.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: use sys_clock implementation

Use the newly added sys_clock API in lib/os for

* clock_gettime()
* clock_settime()
* clock_nanosleep() and nanosleep()
* gettimeofday()

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
libc: common: time: use sys_clock api rather than posix

Remove POSIX clock_gettime() from the common libc time implementation,
since POSIX should not be a dependency for ISO C.

Instead, use the newly added lib/os sys_clock API.

Specifically, sys_clock_gettime().

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
libc: use the common libc time() implementation for most libcs

Use the implementation of time() from the common libc, since there
it no longer pulls in POSIX.

Use is implied for minimal, newlib, and picolibc, and selected
for IAR.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
net: remove dependency on posix for iso c time() function

The ISO C function time() should not depend on POSIX and this was
corrected recently via the common libc time() implementation.

Remove this dependency from the network subsystem where it has
been unduly needed for some time.

Similarly, XSI_SINGLE_PROCESS was a dependency for time() via
picolibc, because the time() implementation in picolibc relies
on the POSIX gettimeofday() call.

However, since Zephyr's common libc time() implementation no longer
depends on that, it can be removed as a dependency in the network
subsystem as well.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
samples: net: remove POSIX_TIMERS and XSI_SINGLE_PROCESS

Remove POSIX_TIMERS and XSI_SINGLE_PROCESS dependencies from the aws
cloud sample and the lwm2m client sample, as they are no longer
required.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
tests: lib: move time testsuite to c_lib

Presumably the time testsuite was separate from the c library set of
testsuites because it had a depedency on POSIX.

Since that dependency no longer exists, colocate the time testsuite
with the other c library testsuites.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
tests: libc: thrd: use timespec_from_timeout()

Use `timespec_from_timeout(K_MSEC(msec), &ts)` instead of leaning on
lazily-crafted timespecs with invalid tv_nsec fields.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
libc: common: thrd: use sys_clock_nanosleep() instead of nanosleep()

Reduce the dependency on POSIX by taking advantage of the newly added
sys_clock_nanosleep().

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
* adding tests for `timespec_to_timeout()` and `timeout_from_timespec()`

I'd like to get those in better shape before making this ready for review.

peter-mitsis commented:
Question ... would the proposed _k_clock_ APIs be better served somewhere under _/subsys_ as something like _sys_clock_ APIs? I am wary of them being placed under _/kernel_.

cfriedt commented:
> Question ... would the proposed _k_clock_ APIs be better served somewhere under _/subsys_ as something like _sys_clock_ APIs? I am wary of them being placed under _/kernel_.

`sys_clock` could probably work, although I think there is a chance that it could start to creep into the kernel as well, if things like per-cpu clocks or thread or process clocks become a thing.

Commit:
tests: libc: thrd: compare with thrd_success rather than ok or zero

The C standard does not guarantee that `thrd_success` is equal to
zero, so ensure that the test takes that into account.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
tests: libc: thrd: do not pass NULL for thrd_sleep() duration

The specification does not say that the first struct timespec passed to
thrd_sleep() may be NULL.

Preserving the testpoint so that a possible future testsuite can be
run that expects faults to occur when undefined behaviour is used.

Faults are one possible solution to handling undefined behaviour, but
it would be good to come to a concensus about how that should be
handled in Zephyr for standard function calls belonging to ISO C or
POSIX.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
doc: release: 4.2.0: add sys_clock gettime settime nanosleep notes

Add bullets to release notes for sys_clock_gettime(),
sys_clock_settime(), and sys_clock_nanosleep().

These changes were significant, as they make timekeeping dependencies
more independent between libc and posix, and remove several unnecessary
dependencies on posix for ISO C library routines.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

peter-mitsis commented on include/zephyr/posix/time.h:
```diff
@@ -60,13 +60,14 @@ struct itimerspec {
 #include <errno.h>
```
Commit message refers to k_clock instead of sys_clock

cfriedt replied:
fixed!

peter-mitsis reviewed (commented):
Looks great to me--spotted one minor remnant from the previous version of the change set in a commit message.

ycsin reviewed (dismissed):
Looks neat 👌

carlescufi commented on include/zephyr/sys/clock.h:
```diff
@@ -0,0 +1,117 @@
+/*
```
I agree with this point. This can be a bit confusing, and given that this is a new API class, would you be ok to discuss this tomorrow in the Architecture WG @cfriedt?

cfriedt replied:
@carlescufi - what do you suggest we do in terms of `sys_clock.h` and `sys/clock.h`? I kind of like the `sys/clock.h` header better (🤷‍♂️ ) but `sys_clock.h` has already existed for a long time.

Would it be sufficient to merge the two files into `sys/clock.h` and have `sys_clock.h` include `sys/clock.h`?

We could at some point deprecate `sys_clock.h`

The alternative is moving `sys/clock.h` contents into `sys_clock.h`.

There seem to be no conflicting toolchain paths that would interfere with `sys/clock.h`.

cfriedt replied:
It seems that simply merging the contents of the two headers will result in compile failures, so I don't think it will be done in this PR.

carlescufi replied:
I personally don't mind either way, as long as we do not introduce this confusion of `sys_clock` vs `sys/clock`. Personally I am in favor of having it use the `sys_` prefix, so for me perhaps the best long-term solution would be to keep a `sys_clock.h` that only includes the new `sys/clock.h` that then has the contents of both.

carlescufi replied:
@cfriedt I still see `sys_clock.h` and `sys/clock.h` with different contents. Can we move the contents of `sys_clock.h` to the new `sys/clock.h`?

carlescufi replied:
Ah, I see you are testing it in a draft PR :) thanks!

cfriedt commented on lib/os/clock.c:
```diff
@@ -0,0 +1,200 @@
+		return -EINVAL;
+	}
+
+	if (!abstime && unlikely(rqtp->tv_sec >= ULLONG_MAX / NSEC_PER_SEC)) {
```
That works - good catch!

cfriedt replied:
Fixed!

cfriedt commented on include/zephyr/sys/clock.h:
```diff
@@ -0,0 +1,117 @@
+ * @brief The monotonic clock
+ *
+ * This steady clock is used to measure time since the system booted. Time from this clock is
+ * always monotonically increasing, modulo the upper bound of `{.tv_sec = INT64_MAX, .tv_nsec =
+ * 1000000000}`.
```
The only architectures supported by Zephyr that do not use a 64-bit `time_t` are `native_sim/native` and `unit_testing`. Due to the `-m64` switch, `native_sim/native/64` uses a 64-bit `time_t`.

There are many good reasons for requiring a 64-bit `time_t`, but the only reason I can think of that the two platforms above do not use a 64-bit `time_t` is because `__USE_TIME_BITS64` is not set. However, that is not a universal macro, so ... #90029.

Maybe it would Just Work to use `-D__USE_TIME_BITS64` when building for `native_sim/native` and `unit_testing`? It's hard to say if that would work for everyone though.

I don't mind removing the comment, in any case.

cfriedt replied:
Fixed!

cfriedt commented:
* used `us` in final `k_sleep()` call in `sys_clock_nanosleep()` for 32-bit compatibility (which is traditionally how the code was structured)
* used `UINT64_MAX` instead of `ULLONG_MAX`
* adjusted doxygen for `SYS_CLOCK_MONOTONIC` to remove the modulo bit
* added syscall `sys_clock_getrtbase()` and made `sys_clock_gettime()` a regular function
  * preserves the micro-optimization that has existed since 2023
  * i.e. avoid a syscall to get rt clock base when the monotonic clock is queried

maass-hamburg commented:
could likely also be changed for `CONFIG_LOG_TIMESTAMP_USE_REALTIME` and it's parts in `logging/log_core.c`

aescolar commented on lib/os/clock.c:
```diff
@@ -0,0 +1,200 @@
+	}
+
+	do {
+		ns = k_sleep(K_TIMEOUT_ABS_NS(ns)) * NSEC_PER_MSEC;
```
Unfortunately it still depends on CONFIG_TIMEOUT_64BIT=y and will fail to build with it set to n.

cfriedt replied:
I'll take a closer look.

Commit:
posix: cond + mutex: log messages generate warnings for 32-bit timeouts

Previously, log messages would generate warnings (escalated to errors
in CI) when CONFIG_TIMEOUT_64BIT=n.

For example,

```
west build -p auto -b qemu_cortex_m0 -t run tests/posix/timers/ \
  -- -DCONFIG_TIMEOUT_64BIT=n
..
warning: format '%llx' expects argument of type \
  'long long unsigned int', but argument 3 has type 'k_ticks_t' \
  {aka 'unsigned int'}
```

Use portable print specifiers and cast the argument to int64_t.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
logging: use sys_clock_gettime()

Use the newly added sys_clock_gettime() to avoid the unnecessary
dependency on POSIX APIs.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
* refactored `z_impl_sys_clock_settime()` to use `timespec_from_ticks()`

maass-hamburg reviewed (dismissed):
LGTM

cfriedt commented on lib/libc/common/source/time/time.c:
```diff
@@ -1,22 +1,23 @@
 /*
  * Copyright (c) 2021 Golioth, Inc.
+ * Copyright (c) 2025 Tenstorrent AI ULC
```
Sorry attributing changes here is not to your liking.

Of the 14 lines of code in this file, this change modifies over 30% of them, which is significant.

Maybe someone from Golioth who holds the current copyright can voice their opinion.

E.g. @beriberikix

andyross reviewed (commented):
Some late-arriving nitpicks.  No major complaints, though I do think some clarification on the rtbase feature would  be good if it's going to be a public API

cfriedt commented on lib/os/clock.c:
```diff
@@ -0,0 +1,222 @@
+		sys_clock_getrtbase(&base);
+		if (!timespec_add(ts, &base)) {
+			return -EOVERFLOW;
+		}
```
It is unfortunately backwards but that's because of the constness of the inputs.

cfriedt replied:
Yeah, I think there are definite improvements to make in this area, and possibly some misnomers. The overall contract is still valid, but quality can definitely be improved over time.

For example, the monotonic clock is definitely cooked as it is, but it's cooked in such a way that it will never give sub-tick clock resolution. I think a long time ago, it was based on cycles, but there were some issues encountered.

On Linux, for example, via the VDSO, the monotonic clock has a resolution of 1ns. This is also implicitly high precision. However, the `_COARSE` clock only has a resolution of about 1ms by default (the tick rate) and naturally the precison is also limited.

https://docs.redhat.com/en/documentation/red_hat_enterprise_linux_for_real_time/7/html/reference_guide/sect-posix_clocks

So our `SYS_CLOCK_MONOTONIC` here is efffectively a `SYS_CLOCK_MONOTONIC_COARSE`.

Definitely agree. I'm less concerned with fixing all aspects of the clock algorithm, and naming private details correctly in this PR and more concerned with the mechanical 'what public symbol goes where', removing unnecessary dependencies, and making sure tests pass.

cfriedt replied:
I think the naming part of this is fixed?

cfriedt commented on include/zephyr/sys/clock.h:
```diff
@@ -0,0 +1,117 @@
+#endif
+
+/**
+ * @addtogroup clock_apis
```
Given there are a few RFCs on this topic already, I think we can skip making a new one.

cfriedt commented on include/zephyr/sys/clock.h:
```diff
@@ -0,0 +1,133 @@
+ *
+ * @param tp Pointer to memory where time will be written.
+ */
+__syscall void sys_clock_getrtbase(struct timespec *tp);
```
Fixed!

cfriedt commented:
* renamed "base" to "offset"
* made `clock_gettime()` saturate rather than report an error when time overflows 292 billion years from now

Commit:
sys: clock: merge sys_clock.h into sys/clock.h

Merge sys_clock.h into sys/clock.h and preserve the original header.

This allows us to use a slightly more intuitive name for the
sys/clock.h header, while reducing churn throughout the codebase.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
> something is failing in ci

Yeah, it's unrelated. I think it was fixed already in main. I likely have one or two more changes to make here to get sys_clock.h and sys/clock.h merged

Commit:
tests: libc: thrd: use sys_clock_gettime() and timespec_add()

Use sys_clock_gettime() instead of clock_gettime() and use
timespec_add() instead of manually adding timespec structures.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
scripts: ci: check_compliance: allow CONFIG_BOARD_UNIT_TESTING

Previously, code like the snippet below would not pass compliance checks.
With this change, we add `CONFIG_BOARD_UNIT_TESTING` to the Kconfig allow
list, so that code like the snippet below is not incorrectly reported as
a compliance failure.

```
if defined(CONFIG_BOARD_UNIT_TESTING)
/* unit-testing only */
endif
```

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
* worked-around / merged CI issues in separate PR's #91616, #91300, and #91615 (c/o @nashif)
* merged `sys_clock.h` contents into `sys/clock.h` (but kept `sys_clock.h` to avoid churn)

cfriedt commented:
@ycsin @maass-hamburg - please revisit

ycsin commented on include/zephyr/sys_clock.h:
```diff
@@ -4,344 +4,9 @@
- * components that use timer functionality.
- */
-
 #ifndef ZEPHYR_INCLUDE_SYS_CLOCK_H_
```
Should we rename this to something else (i.e. `ZEPHYR_INCLUDE_SYSCLOCK_H_`) so that the `sys/clock.h` can take the ownership of `ZEPHYR_INCLUDE_SYS_CLOCK_H_` and be consistent with everything else inside the `include/zephyr/sys/` folder?

ycsin reviewed (dismissed):
+1 to untangling the dependencies

maass-hamburg reviewed (dismissed):
Generally fine to me

ycsin commented on include/zephyr/sys_clock.h:
```diff
@@ -4,344 +4,4 @@
-#endif
-
-#endif /* ZEPHYR_INCLUDE_SYS_CLOCK_H_ */
+#include <zephyr/sys/clock.h>
```
Are we going to deprecate this header in another PR, given that this PR is quite large already?

cfriedt replied:
In a future PR, yeah. It will probably have some deprecation process.

cfriedt commented:
@nashif - this still requires assignee approval
